src/wire/mld.rs

```diff
@@ -292,8 +294,99 @@ impl<T: AsRef<[u8]> + AsMut<[u8]>> AddressRecord<T> {
     }
 }
 
+/// A high-level representation of an MLDv2 packet header.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum Repr<'a> {
+    Query {
+        max_resp_code: u16,
+        mcast_addr: Ipv6Address,
+        s_flag: bool,
+        qrv: u8,
+        qqic: u8,
+        num_srcs: u16,
+        data: &'a [u8]
+    },
+    Report {
+        nr_mcast_addr_rcrds: u16,
+        data: &'a [u8]
+    }
+}
+
+impl<'a> Repr<'a> {
+    /// Parse an MLDv2 packet and return a high-level representation.
+    pub fn parse<T>(packet: &Packet<&'a T>) -> Result<Repr<'a>>
+            where T: AsRef<[u8]> + ?Sized {
+        match packet.msg_type() {
+            Message::MldQuery => {
+                Ok(Repr::Query {
+                    max_resp_code: packet.max_resp_code(),
+                    mcast_addr: packet.mcast_addr(),
+                    s_flag: packet.s_flag(),
+                    qrv: packet.qrv(),
+                    qqic: packet.qqic(),
+                    num_srcs: packet.num_srcs(),
+                    data: packet.payload()
+                })
+            },
+            Message::MldReport => {
+                Ok(Repr::Report {
+                    nr_mcast_addr_rcrds: packet.nr_mcast_addr_rcrds(),
+                    data: packet.payload()
+                })
+            },
+            _ => Err(Error::Unrecognized)
+        }
+    }
+
+    /// Return the length of a packet that will be emitted from this high-level representation.
+    pub fn buffer_len(&self) -> usize {
+        match self {
+            Repr::Query { num_srcs, .. } => {
+                field::QUERY_NUM_SRCS.end + (*num_srcs as usize * mem::size_of::<Ipv6Address>())
+            }
+            Repr::Report { .. } => {
+                // FIXME(dlrobertson): This does not take into account the address records.
+                field::NR_MCAST_RCRDS.end
```

Comment:
The size of each Multicast Address Record is not static. In order to calculate the length of the `MLDv2` header and the contained Address Records, we would need to iterate over the contained Address Records.